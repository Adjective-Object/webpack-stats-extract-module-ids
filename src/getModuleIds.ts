import { Stats as WebpackStats } from 'webpack';

export interface AbbreviatedModule {
    name: string;
}

export type ModuleIdMap = {
    [id: string]: AbbreviatedModule;
};

/**
 * Builds a copy of a module graph tracking child information & with a unique ID associated
 * with each node in the graph.
 *
 * @param graph
 */
export function getModuleIds(stats: WebpackStats.ToJsonOutput): ModuleIdMap {
    if (!stats.modules) {
        throw new Error('no modules in provided webpack stats object');
    }

    const moduleMap: ModuleIdMap = {};
    for (let module of stats.modules) {
        moduleMap[module.id] = {
            name: module.name
        };
    }

    return moduleMap;
}
