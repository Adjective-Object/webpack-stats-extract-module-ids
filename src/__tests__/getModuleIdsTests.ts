import { Stats as WebpackStats } from 'webpack';
import { getModuleIds } from '../getModuleIds';
import { readFile } from 'mz/fs';
import { join as joinPath } from 'path';

type RecursivePartial<T> = {
    [P in keyof T]?: RecursivePartial<T[P]>;
};

describe('getModuleIds', () => {
    it('replicates reason relationships as reasonChildren', () => {
        const testStats: RecursivePartial<WebpackStats.ToJsonOutput> = {
            modules: [
                {
                    name: 'lib/foo',
                    id: 'lib-foo-1'
                },
                {
                    name: 'lib/bar',
                    reasons: [
                        {
                            moduleName: 'lib/foo'
                        }
                    ],
                    id: 'lib-bar-2'
                },
                {
                    name: 'lib/baz',
                    reasons: [
                        {
                            moduleName: 'lib/bar'
                        }
                    ],
                    id: 'lib-baz-3'
                }
            ]
        };

        const outIdMap = getModuleIds(testStats as WebpackStats.ToJsonOutput);

        expect(outIdMap).toEqual({
            'lib-foo-1': {
                name: 'lib/foo'
            },
            'lib-bar-2': {
                name: 'lib/bar'
            },
            'lib-baz-3': {
                name: 'lib/baz'
            }
        });
    });

    it('matches the snapshot', async () => {
        const statsJsonString = await readFile(
            joinPath(__dirname, '/test-data', 'bonsai-stats.json'),
            'utf-8'
        );
        const stats = JSON.parse(statsJsonString);
        const resultGraph = getModuleIds(stats);

        expect(resultGraph).toMatchSnapshot();
    });
});
