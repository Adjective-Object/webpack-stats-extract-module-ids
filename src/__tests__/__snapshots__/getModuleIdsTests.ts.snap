// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getModuleIds matches the snapshot 1`] = `
Object {
  "": Object {
    "name": "css ./node_modules/css-loader/dist/cjs.js??ref--6-oneOf-3-1!./node_modules/postcss-loader/src??postcss!./src/components/stats/ModuleTableBody.css",
  },
  "0": Object {
    "name": "./node_modules/react/index.js",
  },
  "1": Object {
    "name": "./node_modules/@github/octicons-react/dist/index.esm.js",
  },
  "10": Object {
    "name": "./node_modules/invariant/browser.js",
  },
  "11": Object {
    "name": "./node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/esm/defineProperty.js",
  },
  "12": Object {
    "name": "(webpack)/buildin/global.js",
  },
  "13": Object {
    "name": "./node_modules/has/src/index.js",
  },
  "14": Object {
    "name": "./node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/esm/toConsumableArray.js + 3 modules",
  },
  "15": Object {
    "name": "./node_modules/define-properties/index.js",
  },
  "16": Object {
    "name": "./node_modules/function-bind/index.js",
  },
  "17": Object {
    "name": "./node_modules/react-dom/index.js",
  },
  "18": Object {
    "name": "./node_modules/redux/es/index.js + 15 modules",
  },
  "19": Object {
    "name": "./node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/esm/slicedToArray.js + 3 modules",
  },
  "2": Object {
    "name": "./node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/esm/objectSpread.js",
  },
  "20": Object {
    "name": "./node_modules/is-callable/index.js",
  },
  "21": Object {
    "name": "./node_modules/es-abstract/GetIntrinsic.js",
  },
  "22": Object {
    "name": "./node_modules/react-is/index.js",
  },
  "23": Object {
    "name": "./node_modules/hoist-non-react-statics/dist/hoist-non-react-statics.cjs.js",
  },
  "24": Object {
    "name": "./node_modules/symbol-observable/es/index.js",
  },
  "25": Object {
    "name": "./node_modules/object-keys/index.js",
  },
  "26": Object {
    "name": "./node_modules/object-keys/isArguments.js",
  },
  "27": Object {
    "name": "./node_modules/object.values/implementation.js",
  },
  "28": Object {
    "name": "./node_modules/es-abstract/es2015.js",
  },
  "29": Object {
    "name": "./node_modules/es-to-primitive/helpers/isPrimitive.js",
  },
  "3": Object {
    "name": "./node_modules/react-redux/es/index.js + 25 modules",
  },
  "30": Object {
    "name": "./node_modules/es-abstract/helpers/assertRecord.js",
  },
  "31": Object {
    "name": "./node_modules/es-abstract/helpers/isNaN.js",
  },
  "32": Object {
    "name": "./node_modules/es-abstract/helpers/isFinite.js",
  },
  "33": Object {
    "name": "./node_modules/es-abstract/helpers/assign.js",
  },
  "34": Object {
    "name": "./node_modules/es-abstract/helpers/sign.js",
  },
  "35": Object {
    "name": "./node_modules/es-abstract/helpers/mod.js",
  },
  "36": Object {
    "name": "./node_modules/object.values/polyfill.js",
  },
  "37": Object {
    "name": "./node_modules/es-abstract/es6.js",
  },
  "38": Object {
    "name": "./node_modules/array-includes/implementation.js",
  },
  "39": Object {
    "name": "./node_modules/array-includes/polyfill.js",
  },
  "4": Object {
    "name": "./node_modules/prop-types/index.js",
  },
  "40": Object {
    "name": "./node_modules/object-assign/index.js",
  },
  "41": Object {
    "name": "./node_modules/object.values/index.js",
  },
  "42": Object {
    "name": "./node_modules/array-includes/index.js",
  },
  "43": Object {
    "name": "./node_modules/symbol-observable/es/ponyfill.js",
  },
  "44": Object {
    "name": "./node_modules/lodash-es/_freeGlobal.js",
  },
  "45": Object {
    "name": "multi ./src/index.js",
  },
  "46": Object {
    "name": "./node_modules/object-keys/implementation.js",
  },
  "47": Object {
    "name": "./node_modules/es-abstract/es7.js",
  },
  "48": Object {
    "name": "./node_modules/es-abstract/es2016.js",
  },
  "49": Object {
    "name": "./node_modules/function-bind/implementation.js",
  },
  "5": Object {
    "name": "./node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/esm/classCallCheck.js",
  },
  "50": Object {
    "name": "./node_modules/es-to-primitive/es6.js",
  },
  "51": Object {
    "name": "./node_modules/es-to-primitive/es2015.js",
  },
  "52": Object {
    "name": "./node_modules/is-date-object/index.js",
  },
  "53": Object {
    "name": "./node_modules/is-symbol/index.js",
  },
  "54": Object {
    "name": "./node_modules/has-symbols/index.js",
  },
  "55": Object {
    "name": "./node_modules/has-symbols/shams.js",
  },
  "56": Object {
    "name": "./node_modules/es-abstract/helpers/isPrimitive.js",
  },
  "57": Object {
    "name": "./node_modules/es-abstract/es5.js",
  },
  "58": Object {
    "name": "./node_modules/es-to-primitive/es5.js",
  },
  "59": Object {
    "name": "./node_modules/is-regex/index.js",
  },
  "6": Object {
    "name": "./node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/esm/createClass.js",
  },
  "60": Object {
    "name": "./node_modules/object.values/shim.js",
  },
  "61": Object {
    "name": "./node_modules/array-includes/shim.js",
  },
  "62": Object {
    "name": "./node_modules/raf/polyfill.js",
  },
  "63": Object {
    "name": "./node_modules/raf/index.js",
  },
  "64": Object {
    "name": "./node_modules/performance-now/lib/performance-now.js",
  },
  "65": Object {
    "name": "./node_modules/process/browser.js",
  },
  "66": Object {
    "name": "./node_modules/console-shim/console-shim.js",
  },
  "67": Object {
    "name": "./node_modules/react/cjs/react.production.min.js",
  },
  "68": Object {
    "name": "./node_modules/prop-types/factoryWithThrowingShims.js",
  },
  "69": Object {
    "name": "./node_modules/prop-types/lib/ReactPropTypesSecret.js",
  },
  "7": Object {
    "name": "./node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/esm/getPrototypeOf.js",
  },
  "70": Object {
    "name": "./node_modules/react-is/cjs/react-is.production.min.js",
  },
  "71": Object {
    "name": "(webpack)/buildin/harmony-module.js",
  },
  "72": Object {
    "name": "./node_modules/react-dom/cjs/react-dom.production.min.js",
  },
  "73": Object {
    "name": "./node_modules/scheduler/index.js",
  },
  "74": Object {
    "name": "./node_modules/scheduler/cjs/scheduler.production.min.js",
  },
  "75": Object {
    "name": "./src/components/OffsetPageAnchor.css",
  },
  "76": Object {
    "name": "./src/components/stats/ModuleTableBody.css",
  },
  "77": Object {
    "name": "./src/components/LoadingSpinner.css",
  },
  "78": Object {
    "name": "./src/components/stats/ChunkDropdown.css",
  },
  "79": Object {
    "name": "./src/components/App.css",
  },
  "8": Object {
    "name": "./node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/esm/possibleConstructorReturn.js + 2 modules",
  },
  "80": Object {
    "name": "./node_modules/bootstrap/dist/css/bootstrap.min.css",
  },
  "81": Object {
    "name": "./src/index.css",
  },
  "82": Object {
    "name": "./src/index.js + 69 modules",
  },
  "9": Object {
    "name": "./node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/esm/inherits.js + 1 modules",
  },
}
`;
